soroban-env-host/src/crypto/bn254.rs

```diff
@@ -0,0 +1,313 @@
+            }
+            // Split the 64-byte big-endian encoding into 32-byte X and Y limbs.
+            let mut x_be = [0u8; BN254_FP_SERIALIZED_SIZE];
+            let mut y_be = [0u8; BN254_FP_SERIALIZED_SIZE];
+            x_be.copy_from_slice(&bytes[0..32]);
+            y_be.copy_from_slice(&bytes[32..64]);
+
+            // Convert from big-endian to little-endian, which the field deserializer expects.
+            let mut x_le = x_be;
+            let mut y_le = y_be;
+            x_le.reverse();
+            y_le.reverse();
+            // Deserialize coordinates into field elements without validating curve/subgroup membership.
+            let x: Fq = self.bn254_deserialize_unchecked(&x_le, "G1.x")?;
+            let y: Fq = self.bn254_deserialize_unchecked(&y_le, "G1.y")?;
+            // Construct the affine point without checks; on-curve and subgroup checks follow below.
+            Ok(G1Affine::new_unchecked(x, y))
+        })?;
+
+        if !self.bn254_check_point_is_on_curve(&pt)? {
+            return Err(self.bn254_err_invalid_input("bn254 G1: point not on curve"));
+        }
+
+        if subgroup_check && !self.bn254_check_point_is_in_subgroup(&pt)? {
+            return Err(self.bn254_err_invalid_input("bn254 G1: point not in the correct subgroup"));
+        }
+        Ok(pt)
+    }
+
+    pub(crate) fn bn254_g2_affine_deserialize_from_bytesobj(
+        &self,
+        bo: BytesObject,
+        subgroup_check: bool,
+    ) -> Result<G2Affine, HostError> {
+        let pt = self.visit_obj(bo, |bytes: &ScBytes| {
+            if bytes.len() != BN254_G2_SERIALIZED_SIZE {
+                return Err(self.err(
+                    ScErrorType::Crypto,
+                    ScErrorCode::InvalidInput,
+                    "bn254 G2: invalid input length to deserialize",
+                    &[
+                        Val::from_u32(bytes.len() as u32).into(),
+                        Val::from_u32(BN254_G2_SERIALIZED_SIZE as u32).into(),
+                    ],
+                ));
+            }
+            if bytes.iter().all(|b| b.is_zero()) {
+                return Ok(G2Projective::zero().into_affine());
+            }
+            let mut x_c1_be = [0u8; 32];
+            let mut x_c0_be = [0u8; 32];
+            let mut y_c1_be = [0u8; 32];
+            let mut y_c0_be = [0u8; 32];
+
+            x_c1_be.copy_from_slice(&bytes[0..32]);
+            x_c0_be.copy_from_slice(&bytes[32..64]);
+            y_c1_be.copy_from_slice(&bytes[64..96]);
+            y_c0_be.copy_from_slice(&bytes[96..128]);
+
+            let mut x_c1_le = x_c1_be;
+            let mut x_c0_le = x_c0_be;
+            let mut y_c1_le = y_c1_be;
+            let mut y_c0_le = y_c0_be;
+            x_c1_le.reverse();
+            x_c0_le.reverse();
+            y_c1_le.reverse();
+            y_c0_le.reverse();
+
+            let x_c1: Fq = self.bn254_deserialize_unchecked(&x_c1_le, "G2.x.c1")?;
```

Comment:
I noticed that you de-serialized G1 and G2 by each coordinate, but you should be able to de-serialize them as a whole right? Using `deserialize_with_mode`?